Add lean style math symbol completions

johannesCmayer opened the pull request:
Add math symbol completions based on the [VScode lean extension](https://github.com/leanprover/vscode-lean).

AucaCoyan commented:
Hello! The CI found an error
>Validating package: lean-symbols... Found 1 errors:
Check: incoherent_path
 ->: package name mismatch in package lean-symbols: the path indicates that the name is `lean-symbols`, but the manifest calls it `espanso-lean-symbols`
 
 Would you kindly fix it? It's almost there 😄

smeech reviewed (changes requested):
A straightforward text replacement package.
No `cmd:`, `script`, `shell` or risky code.
Could safely be merged.

johannesCmayer commented:
I have now updated the package to incorporate all of the fixes. I will make another pull request because I actually regenerated all completions to the latest completions, and also removed all my custom modifications such that it is actually exactly the lean completions.

My main question is whether it makes sense to put the 30-line Python script that generates the completions in the package. It's not required, and it's not ever used, but it also seems like the natural place to put it.

smeech commented:
> I have now updated the package to incorporate all of the fixes. I will make another pull request because I actually regenerated all completions to the latest completions, and also removed all my custom modifications such that it is actually exactly the lean completions.
> 
> My main question is whether it makes sense to put the 30-line Python script that generates the completions in the package. It's not required, and it's not ever used, but it also seems like the natural place to put it.

That seems reasonable. 30 lines isn't large, it may be of interest, and I suppose it would enable users to update the list if it changed, although you might update the package with a new version if that occurred. 
Include some explanatory text in your new README.md, perhaps.

smeech commented on packages/lean-symbols/1.0.0/abbreviations.json:
```diff
@@ -0,0 +1,1836 @@
+{
```
Do we need this file, as you have provided a link to the source in your README.md (thank you)?

johannesCmayer replied:
We don't strictly need it. I put in, in case the file in the VScode extension is updated and becomes incompatible, but I could of cause also just point to a specific commit. I think I will do that instead.

johannesCmayer replied:
Ah it already does this actually.

johannesCmayer replied:
I have made it such that the file is fetched now and deleted it from the repo.

smeech commented on packages/lean-symbols/1.0.0/README.md:
```diff
@@ -1 +1,6 @@
-Insert Lean Symbols with espanso.
+# Insert the lean unicode math symbols with espanso
```
Thank you - that's much more useful.

smeech commented on packages/lean-symbols/1.0.0/custom_non_overriding_map.yml:
```diff
@@ -0,0 +1 @@
+map: ↦
```
Is this needed, or could it be generated (if not already in existence) by `generate_mapping.py`?

johannesCmayer replied:
That is what `generate_mapping.py` uses as input. Instead of hardcoding it into the code I made it a separate file such that it is easier to modify.

smeech replied:
Sorry - maybe I wasn't clear enough in my question.
As it consists of just a tiny file appearing to contain just `map: ↦`, is there a reason why it couldn't be created on the fly by `generate_mapping.py` at runtime, rather than supplying it as part of the package?
I suspect most people will just install the package and use it, without investigating `generate_mapping.py`, although it will be of interest to a few.
Sorry if this seems picky - I'm just trying to understand!

Thank you for making the other changes. I think it's ready to go otherwise.

johannesCmayer replied:
Well, maybe I am overengineering this. I thought that I might make more changes in future versions, but right now it's only one additional mapping. If I want additional mappings it would be quite easy to add them by just editing this file. Though there is a chance that I will not use it. As a reference for the old version, I think I added over 30 symbols and changed many other mappings. Though I now want to intentionally not change the existing mappings anymore.

If it's really a problem I can move it into `generate_mapping.py`.

smeech replied:
Thank you for the explanation! It's always tricky converting a personal project to something for wider use.

To keep things simple, if it doesn't make a difference, and it's not too much trouble, would you remove `custom_non_overriding_map.yml` and add a few lines to `generate_mapping.py` to create it as required, please? If it's something that needs to persist from one run to another (showing my ignorance of Python yaml coding here!), do add a check whether it exists already, to avoid overwriting it.

johannesCmayer replied:
Ok, wait. Just to make sure let me explain in a bit more detail. This is a configuration file for the `generate_mapping.py` program. When you start the program it downloads the lean mappings and then loads this file and injects the mappings in this file into the lean mappings, and from the combined mappings then generates the espanso mappings.

This is not a file that needs to be generated at runtime. It's a file containing additional data that configures what the final mappings will look like. Maybe I failed to explain this properly before.

The alternative is to not have a configurable file but to HARDCODE into the code that this mappings should exist. I thought that was a bit less elegant. So that is the change that I would make. Hardcode the data in the script file.

Sorry if this was all clear and I am overexplaining things.

smeech replied:
That's helpful. It's such a small file that I thought it could just be created when required.

I'll leave it to you - just let me know when you're ready to go.

johannesCmayer replied:
From my side, I'd prefer to keep the current structure of the file such that it is easier to add additional mappings (which might be shortcuts to existing mapping, that is what the one mapping in the file currently is doing. Also funnily enough that shortcut is map which is a shorthand for mapsto which maps to the symbol '↦' that denotes how a mathematical function maps an input to an output).

smeech reviewed (commented):
Thank you - this is coming together nicely.
Just a couple of new comments - sorry :frowning_face: !

smeech commented on packages/lean-symbols/1.0.0/_manifest.yml:
```diff
@@ -0,0 +1,10 @@
+  extension allows you to insert, using the same shortcuts. For example '\a\', '\a ', '\aTAB', 
+  all insert α, i.e. the greek alpha. TAB denotes a single press of the tab key.
+
+version: 0.1.0
```
Fails quality check - the path indicates that the version is 1.0.0, but the manifest indicates version 0.1.0.

johannesCmayer replied:
I fixed that now.
